main.go

```diff
@@ -453,25 +505,220 @@ func (c *controller) waitForCacheSync(stop <-chan struct{}) error {
 	return nil
 }
 
-func (c *controller) worker() {
-	for c.queue.get() {
+func (c *controller) podWorker() {
+	fn := func() bool {
+		key, quit := c.podQ.Get()
+		if quit {
+			return false
+		}
+		defer c.podQ.Done(key)
+		if err := c.cleanUp(key.(*corev1.Pod)); err != nil {
+			level.Error(c.logger).Log("msg", "unable to clean up PVC", "err", err)
+			c.podQ.AddRateLimited(key)
+			return true
+		}
+		c.podQ.Forget(key)
+		return true
+	}
+	for fn() {
+	}
+}
+
+func (c *controller) stsWorker() {
+	for c.stsQ.get() {
 		c.sync()
 	}
 }
 
+func (c *controller) resolvePodOwnerRef(namespace string, refs []metav1.OwnerReference) (*appsv1.StatefulSet, error) {
+	for _, ref := range refs {
+		// If the owner reference points at the wrong kind of object, skip.
+		if ref.Kind != "StatefulSet" {
+			continue
+		}
+		// If the owner reference points at something that we don't have, then skip.
+		obj, ok, err := c.ssetInf.GetStore().GetByKey(fmt.Sprintf("%s/%s", namespace, ref.Name))
+		if !ok {
+			continue
+		}
+		if err != nil {
+			return nil, err
+		}
+		sts := obj.(*appsv1.StatefulSet)
+		if sts.UID != ref.UID {
+			return nil, errors.Wrap(err, "owner reference UID does not match StatefulSet")
+		}
+		return sts, nil
+	}
+	return nil, nil
+}
+
+func (c *controller) generateHelper(name string, sts *appsv1.StatefulSet) *corev1.Pod {
+	initContainerTemplate := corev1.Container{
+		Name:            "replicate",
+		Image:           c.options.cleanupImage,
+		ImagePullPolicy: corev1.PullIfNotPresent,
+		Args: []string{
+			"run",
+			"--single-run",
+			"--objstorefrom.config=$(OBJSTORE_CONFIG_FROM)",
+			"--objstoreto.config=$(OBJSTORE_CONFIG_TO)",
+			"--log.level=debug",
+		},
+		Env: []corev1.EnvVar{
+			{
+				Name:  "OBJSTORE_CONFIG_FROM",
+				Value: "type: FILESYSTEM\nconfig:\n\tdirectory: " + fromMountPath,
+			},
+			{
+				Name: "OBJSTORE_CONFIG_TO",
+				ValueFrom: &corev1.EnvVarSource{
+					SecretKeyRef: &corev1.SecretKeySelector{
+						LocalObjectReference: corev1.LocalObjectReference{
+							Name: sts.Labels[objstoreSecretLabelKey],
+						},
+						Key: sts.Labels[objstoreSecretKeyLabelKey],
+					},
+				},
+			},
+		},
+	}
+	// Inject extra environment variables into the cleanup Pod if provided.
```

Comment:
what is this meant for?

Reply:
so we can inject e.g. AWS_SECRET_ACCESS_KEY AWS_ACCESS_KEY_ID, which we don't put inside of the thanos objstore secret

Reply:
Hmm .. I feel like we should read the job/pod template from "disk" and mount it into the controller via a configmap, there will be endless customizations people will want to do.

Reply:
I considered this but I am a bit hesitant. I would really want to keep the scope of customizations minimal, especially because there controller has to override several fields:
* volumes (to specify the pvcs)
* container volume mounts ( to mount the pvcs)
* container env (to set the 2 objstore configs, one of which points to the local pvc)
* container cmd + args (to pass the env var to the cmd)
In order to override those fields we need to select the container index in a stable way.